File: backend/src/main/java/com/example/backend/model/Booking.java
package com.example.backend.model;

import com.example.backend.common.BookingStatus;
import jakarta.persistence.*;
import lombok.AllArgsConstructor;
import lombok.Builder;
import lombok.Data;
import lombok.NoArgsConstructor;
import org.hibernate.annotations.CreationTimestamp;
import org.hibernate.annotations.UpdateTimestamp;

import java.time.LocalDate;
import java.time.LocalDateTime;
import java.util.List;
import java.util.UUID;

@Entity
@Table(name = "bookings")
@Data
@NoArgsConstructor
@AllArgsConstructor
@Builder
public class Booking {
    @Id
    @GeneratedValue(strategy = GenerationType.UUID)
    @Column(name = "booking_id")
    private UUID id;

    @Column(name = "check_in_date", nullable = false)
    private LocalDate checkInDate;

    @Column(name = "check_out_date", nullable = false)
    private LocalDate checkOutDate;

    @Column(name = "total_amount", nullable = false)
    private Double totalAmount;

    @Enumerated(EnumType.STRING)
    @Column(name = "status", nullable = false, length = 50)
    private BookingStatus status;

    @Column(name = "confirmation_code", nullable = false, unique = true, length = 20)
    private String confirmationCode;

    @Column(name = "guest_count")
    private Integer guestCount;

    @Column(name = "notes", columnDefinition = "TEXT")
    private String notes;

    @Builder.Default
    @Column(name = "is_active")
    private Boolean isActive = true;

    @CreationTimestamp
    @Column(name = "created_at", updatable = false)
    private LocalDateTime createdAt;

    @UpdateTimestamp
    @Column(name = "updated_at")
    private LocalDateTime updatedAt;

    // N:1 Relationship with Hotel
    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "hotel_id", nullable = false)
    private Hotel hotel;

    // N:1 Relationship with Promotion
    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "promo_id")
    private Promotion promotion;

    // N:1 Relationship with User
    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "user_id")
    private User user;

    // 1: N Relationship with BookingRoom
    @OneToMany(mappedBy = "booking", cascade = CascadeType.ALL, orphanRemoval = true)
    private List<BookingRoom> bookingRooms;

    // 1:1 Relationship with Transactions
    @OneToOne(mappedBy = "booking", cascade = CascadeType.ALL, orphanRemoval = true)
    private Transaction transaction;
}

File: frontend/src/types/booking.ts
import type { BaseEntity, BaseFilterRequest } from "./common";
import type { BookingRoomRequest, BookingRoomResponse } from "./bookingRoom";
import type { BookingStatus } from "./enum";
import type { PromotionResponse } from "./promotion";
import type { HotelListResponse } from "./hotel";
import type { UserListResponse } from "./user";

export interface BookingResponse extends BaseEntity {
  checkInDate: string;
  checkOutDate: string;
  totalAmount: number;
  status: BookingStatus;
  confirmationCode: string;
  guestCount: number;
  notes: string;
  hotel: HotelListResponse;
  user: UserListResponse;
  promotion: PromotionResponse;
  bookingRooms: BookingRoomResponse[];
}

export interface BookingListResponse extends BaseEntity {
  checkInDate: string;
  checkOutDate: string;
  totalAmount: number;
  status: BookingStatus;
  confirmationCode: string;
  guestCount: number;
  hotelName: string;
  notes: string;
}

export interface BookingRequest {
  hotelId: string;
  checkInDate: string;
  checkOutDate: string;
  guestCount: number;
  bookingRooms: BookingRoomRequest[];
  notes?: string;
  promoCode?: string;
}

export interface BookingFilterRequest extends BaseFilterRequest {
  userId?: string;
  hotelId?: string;
  promotionId?: string;
  status?: BookingStatus;
  checkInFrom?: string;
  checkInTo?: string;
  checkOutFrom?: string;
  checkOutTo?: string;
  minTotalAmount?: number;
  maxTotalAmount?: number;
}
